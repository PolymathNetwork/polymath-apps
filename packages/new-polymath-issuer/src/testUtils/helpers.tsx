import React, { Fragment } from 'react';
// import createSagaMiddleware from 'redux-saga';
import { render } from 'react-testing-library';
import configureMockStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { Action, Middleware } from 'redux';
import { set } from 'lodash';
import Web3FakeProvider from 'web3-fake-provider';
import { RootState } from '~/state/reducers/root';

interface RenderTestOpts {
  initialState: RootState;
}

const middlewares: Middleware[] = [];
const mockStore = configureMockStore(middlewares);

const customRender = (
  node: React.ReactElement<any>,
  testOpts: RenderTestOpts,
  ...opts: any[]
) => {
  const initialState = testOpts.initialState;
  const { rerender, ...result } = render(
    <Provider store={mockStore(initialState)}>{node}</Provider>,
    ...opts
  );

  const customRerender = (rerenderNode: React.ReactElement<any>) => {
    return rerender(<Fragment>{rerenderNode}</Fragment>);
  };

  return { rerender: customRerender, ...result };
};

export * from 'react-testing-library';
export { customRender as render };

export class MockedStore {
  public state: RootState;
  public dispatched: Action[] = [];

  constructor() {
    const { reducer: rootReducer } = require('~/state/reducers/root');
    this.state = rootReducer(undefined, {} as any);
    this.dispatched = [];
  }

  public dispatch = (action: Action) => {
    const { reducer: rootReducer } = require('~/state/reducers/root');
    this.state = rootReducer(this.getState(), action);
    this.dispatched.push(action);
  };

  public getState = () => {
    return { ...this.state };
  };

  public setState = (path: string, value: any) => {
    const newState = {
      ...this.state,
    };
    set(newState, path, value);
    this.state = newState;
  };
}

export class MockEthereumProvider extends Web3FakeProvider {
  public networkVersion = '15';
  public async enable() {}
}

export function getGeneratorOutputs(gen: IterableIterator<any>) {
  const results = [];
  let next = gen.next();
  while (!next.done) {
    results.push(next.value);
    next = gen.next();
  }
  return results;
}

export function mockEthereumBrowser() {
  const ethereum = new MockEthereumProvider();
  (global as any).ethereum = ethereum;
}
