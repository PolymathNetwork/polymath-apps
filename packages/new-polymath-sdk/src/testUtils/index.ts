import Web3PromiEvent from 'web3-core-promievent';
import { GenericContract } from '~/LowLevel/types';
import { PostTransactionResolver } from '~/PostTransactionResolver';

export class MockedContract<T extends GenericContract> {
  public autoResolve: boolean;
  public errorMsg?: string;
  public fakeTxOnePromiEvent = new Web3PromiEvent();
  public fakeTxTwoPromiEvent = new Web3PromiEvent();
  public failureTxPromiEvent = new Web3PromiEvent();

  public fakeTxOne = jest.fn(async () => {
    return () => {
      if (this.autoResolve) {
        this.fakeTxOnePromiEvent.resolve();
      }
      return this.fakeTxOnePromiEvent.eventEmitter;
    };
  });
  public fakeTxTwo = jest.fn(async () => {
    return () => {
      if (this.autoResolve) {
        this.fakeTxTwoPromiEvent.resolve();
      }
      return this.fakeTxTwoPromiEvent.eventEmitter;
    };
  });

  public failureTx = jest.fn(async () => {
    return () => {
      const err = this.errorMsg || 'Test error';
      this.failureTxPromiEvent.reject(new Error(err));
      return this.failureTxPromiEvent.eventEmitter;
    };
  });

  constructor({
    autoResolve = true,
    errorMsg,
  }: { autoResolve?: boolean; errorMsg?: string } = {}) {
    this.autoResolve = autoResolve;
    this.errorMsg = errorMsg;
  }
}

export const getMockTransactionSpec = (
  method: (args: any) => Promise<any>,
  args: any,
  resolver = async () => {}
) => ({
  method,
  args,
  postTransactionResolver: new PostTransactionResolver(resolver),
});
