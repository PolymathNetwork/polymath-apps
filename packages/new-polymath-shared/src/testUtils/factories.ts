import {
  ProcedureTypes,
  TransactionQueueStatus,
  TransactionQueuePojo,
} from '../typing/types';

export const transactionQueuePojo = (
  data?: Partial<TransactionQueuePojo>
): TransactionQueuePojo => {
  return {
    transactions: [],
    status: TransactionQueueStatus.Idle,
    procedureType: ProcedureTypes.Approve,
    args: {},
    uid: 'fakeUID',
    ...(data || {}),
  };
};
