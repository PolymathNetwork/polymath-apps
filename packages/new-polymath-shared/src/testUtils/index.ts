export * from './factories';

const originalWindow = {
  ...window,
};

interface WindowWithEthereum extends Window {
  ethereum: {
    networkVersion: string;
    _metamask?: {
      isApproved: () => Promise<boolean>;
    };
    enable(): Promise<any>;
  };
}

interface WindowWithWeb3 extends Window {
  web3: {
    version: string;
  };
}

interface MockEthereumBrowserArgs {
  support?: 'legacy' | 'modern';
  options?: {
    networkId: number;
    loaded: boolean;
  };
}

export function mockEthereumBrowser({
  support = 'modern',
  options = {
    networkId: 1,
    loaded: true,
  },
}: MockEthereumBrowserArgs) {
  const { networkId, loaded } = options;

  if (support === 'modern') {
    window = (Object.assign(window, {
      ethereum: {
        networkVersion: loaded ? `${networkId}` : undefined,
      },
    }) as any) as WindowWithEthereum;
  }
  if (support === 'legacy') {
    window = Object.assign(window, {
      web3: {
        version: networkId,
      },
    } as any) as WindowWithWeb3;
  }

  return {
    restore: () => {
      window = originalWindow;
    },
    load: () => {
      if (support === 'modern') {
        (window as WindowWithEthereum).ethereum.networkVersion = `${networkId}`;
      }
      if (support === 'legacy') {
        (window as WindowWithWeb3).web3.version = `${networkId}`;
      }
    },
  };
}
