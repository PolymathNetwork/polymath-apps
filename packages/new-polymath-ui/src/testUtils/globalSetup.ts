import { cleanup } from 'react-testing-library';

afterEach(cleanup);
