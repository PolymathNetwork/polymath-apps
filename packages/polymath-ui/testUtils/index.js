import React from 'react';
import { render } from 'react-testing-library';
import { ThemeProvider } from 'styled-components';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import theme from '../src/theme';

const middlewares = [thunk];

const customRender = (node, ...opts) => {
  const { rerender, ...result } = render(
    <ThemeProvider theme={theme}>{node}</ThemeProvider>,
    ...opts
  );

  const customRerender = (rerenderNode, ...rerenderOpts) => {
    return rerender(
      <ThemeProvider theme={theme}>{rerenderNode}</ThemeProvider>,
      ...rerenderOpts
    );
  };

  return { rerender: customRerender, ...result };
};

export const mockStore = configureMockStore(middlewares);
export * from 'react-testing-library';
export { customRender as render };
