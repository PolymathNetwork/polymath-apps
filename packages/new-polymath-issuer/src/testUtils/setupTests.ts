import '@babel/polyfill';
import dotenv from 'dotenv';
import { cleanup } from 'react-testing-library';
import { mockEthereumBrowser } from './helpers';

// Avoid breaking tests trying to import third party css
jest.mock('typeface-overpass', () => ({}));
jest.mock('simplebar/dist/simplebar.css', () => ({}));

const res = dotenv.config({ path: './config/.env.tests' });

// Override environment variables since we are in test mode and need to isolate them
process.env = {
  ...process.env,
  ...res.parsed,
};

mockEthereumBrowser();

afterEach(cleanup);
