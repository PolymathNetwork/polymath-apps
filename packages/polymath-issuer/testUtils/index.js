import thunk from 'redux-thunk';
import configureMockStore from 'redux-mock-store';

import * as factories from './factories';

const middlewares = [thunk];
export const mockStore = configureMockStore(middlewares);

export { factories };
