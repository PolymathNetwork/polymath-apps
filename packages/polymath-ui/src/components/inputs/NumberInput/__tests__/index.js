import React from 'react';
import { render, fireEvent } from '../../../../../testUtils';
import { NumberInput } from '../index';

let defaultProps;

/**
 * Props for primitive:
 * - name
 * - value
 * - onChange
 * - onBlur
 *
 * Props for interface to Formik
 * - field { name, value }
 * - form { setFieldValue, errors, touched }
 */

// TODO @RafaelVidaurre: Support decimals config, length limit, negative values
describe('NumberInput', () => {
  test('renders without crashing ', () => {
    const { container } = render(<NumberInput name="foo" value={12344.555} />);

    expect(container.firstChild).toMatchSnapshot();
  });

  test('updates display value when field value changes', () => {
    const { container, rerender } = render(<NumberInput />);

    rerender(<NumberInput name="foo" value={123.456} />);
    expect(container.querySelector('input').value).toEqual('123.456');
  });

  test('calls onChange with new numeric value when display value changes to a final state', () => {
    // A final state is any state where a number value can be parsed Examples:
    // "0.1", "123" are final
    // "0.0", "12." are not final
    const onChange = jest.fn();
    const { container } = render(
      <NumberInput value={1} name="foo" onChange={onChange} />
    );
    const input = container.getByTestId('base-input');
    fireEvent.change(input, { target: { value: '12' } });
    expect(onChange).toHaveBeenCalledWith(12);
  });

  test('does not call onChange on render', () => {
    const onChange = jest.fn();
    render(<NumberInput value={1} name="foo" onChange={onChange} />);
    expect(onChange).not.toHaveBeenCalled();
  });

  test('does not update field value if the input is in an intermediate state', () => {
    // Valid states where we cannot yet get a new value update
    // Examples:
    // "." could become ".1"
    // "0.12000" could become "0.120003"
    const onChange = jest.fn();
    const { container } = render(
      <NumberInput value={1} name="foo" onChange={onChange} />
    );
    const input = container.getByTestId('base-input');
    fireEvent.change(input, { target: { value: '1.' } });
    expect(onChange).not.toHaveBeenCalled();
  });

  test('prepends a "0" if the display value starts with a "." ', () => {
    const { container } = render(<NumberInput value={1} name="foo" />);
    const input = container.getByTestId('base-input');
    fireEvent.change(input, { target: { value: '.' } });
    expect(input.value).toEqual('0.');
    fireEvent.change(input, { target: { value: '.123' } });
    expect(input.value).toEqual('0.123');
  });

  test('removes prepended "0" from display value unless it is in first digits position', () => {
    // Examples: "01", "04.1234"
    const { container } = render(<NumberInput value={0} name="foo" />);
    const input = container.getByTestId('base-input');
    fireEvent.change(input, { target: { value: '00' } });
    expect(input.value).toEqual('0');
    fireEvent.change(input, { target: { value: '00.123' } });
    expect(input.value).toEqual('0.123');
    fireEvent.change(input, { target: { value: '001444' } });
    expect(input.value).toEqual('1444');
  });

  test('does not allow changes that reach a non-intermediate invalid state', () => {
    // Examples of these states are "0.." "00"
    const { container } = render(<NumberInput value={0} name="foo" />);
    const input = container.getByTestId('base-input');
    // Intermediate state
    fireEvent.change(input, { target: { value: '0.' } });
    // Invalid state
    fireEvent.change(input, { target: { value: '0..' } });
    expect(input.value).toEqual('0.');
  });

  test('calls onBlur when the input looses focus', () => {
    const onBlur = jest.fn();
    const { container } = render(
      <NumberInput value={0} name="foo" onBlur={onBlur} />
    );
    const input = container.getByTestId('base-input');
    fireEvent.blur(input);
    expect(onBlur).toHaveBeenCalled();
  });
});

describe('NumberInputField', () => {
  test('pending', () => {});
});
