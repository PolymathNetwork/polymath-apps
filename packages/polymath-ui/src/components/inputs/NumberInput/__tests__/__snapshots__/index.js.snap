// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NumberInput renders without crashing  1`] = `
<div
  class="sc-bdVaJa dVtwCI"
>
  <input
    class="sc-bxivhb kRpXks sc-bwzfXH csQacs"
    form="[object Object]"
    id="foo"
    name="foo"
    type="text"
    value=""
  />
</div>
`;
